tests/integration/backends/test_tpp.py

```diff
@@ -471,6 +472,163 @@ def test_clinical_events(select_all_tpp):
+            LowerBound=1,
+            UpperBound=2,
+        ),
+        CodedEventRange(
+            Patient_ID=1,
+            CodedEvent_ID=2,
+            Comparator=4,
+            LowerBound=2,
+            UpperBound=3,
+        ),
+        CodedEventRange(
+            Patient_ID=1,
+            CodedEvent_ID=3,
+            Comparator=5,
+            LowerBound=3,
+            UpperBound=4,
+        ),
+        CodedEventRange(
+            Patient_ID=1,
+            CodedEvent_ID=4,
+            Comparator=6,
+            LowerBound=4,
+            UpperBound=5,
+        ),
+        CodedEventRange(
+            Patient_ID=1,
+            CodedEvent_ID=5,
+            Comparator=7,
+            LowerBound=5,
+            UpperBound=6,
+        ),
+        CodedEventRange(
+            Patient_ID=1,
+            CodedEvent_ID=6,
+            Comparator=8,
+            LowerBound=6,
+            UpperBound=7,
+        ),
+    )
+    expected = [
+        {
+            "patient_id": 1,
+            "date": date(2020, 10, 20),
+            "snomedct_code": None,
+            "ctv3_code": "xyz",
+            "numeric_value": None,
+            "comparator": "~",
+            "lower_bound": 1.0,
+            "upper_bound": 2.0,
+        },
+        {
+            "patient_id": 1,
+            "date": date(2020, 11, 21),
+            "snomedct_code": "ijk",
+            "ctv3_code": None,
+            "numeric_value": 1.5,
+            "comparator": "=",
+            "lower_bound": 2.0,
+            "upper_bound": 3.0,
+        },
+        {
+            "patient_id": 1,
+            "date": date(2020, 12, 20),
+            "snomedct_code": None,
+            "ctv3_code": "xyz",
+            "numeric_value": None,
+            "comparator": ">=",
+            "lower_bound": 3.0,
+            "upper_bound": 4.0,
+        },
+        {
+            "patient_id": 1,
+            "date": date(2021, 1, 21),
+            "snomedct_code": "ijk",
+            "ctv3_code": None,
+            "numeric_value": 1.5,
+            "comparator": ">",
+            "lower_bound": 4.0,
+            "upper_bound": 5.0,
+        },
+        {
+            "patient_id": 1,
+            "date": date(2021, 2, 20),
+            "snomedct_code": None,
+            "ctv3_code": "xyz",
+            "numeric_value": None,
+            "comparator": "<",
+            "lower_bound": 5.0,
+            "upper_bound": 6.0,
+        },
+        {
+            "patient_id": 1,
+            "date": date(2021, 3, 21),
+            "snomedct_code": "ijk",
+            "ctv3_code": None,
+            "numeric_value": 1.5,
+            "comparator": "<=",
+            "lower_bound": 6.0,
+            "upper_bound": 7.0,
+        },
+    ]
+    # cannot rely on input order equalling output order,
+    # dicts are neither hashable nor sortable
+    assert not [r for r in results if r not in expected]
+    assert not [e for e in expected if e not in results]
```

Comment:
It tests that all the comparator mappings in the CASE statement are correct (assuming they are correctly specified here), based on recent experience that feels like a good idea. Perhaps could be more concise.

The reason for the ordering is that the CTv3 codes come first because that table is first in the UNION. I guess I could just re-order the expected list to match that